src/crimson/net/SocketMessenger.cc

```diff
@@ -134,6 +134,9 @@ seastar::future<> SocketMessenger::start(Dispatcher *disp) {
 crimson::net::ConnectionRef
 SocketMessenger::connect(const entity_addr_t& peer_addr, const entity_name_t& peer_name)
 {
+  if (stopping)
```

Comment:
is it possible to ensure the shutdown process start with the inner most components? something like

1. pg
1. osd
1. osd infras
   1. mgr client
   1. mon client
1. msgr
1. configs, etc.

?

so nobody will try to establish a connection when messenger is stopping?

Reply:
> so nobody will try to establish a connection when messenger is stopping?

I think this implies that each component needs to implement its own `stop()` method properly, to make sure that all the internal continuations/tasks are finished when the future is fulfilled. And it also needs to refuse to process further events from components that are still not shutdown yet. For example, messenger doesn't know that mon client is stopped and will keep dispatching events to it, that's why there are cumbersome `handle_exception_type(...)` logic in this patch.

Reply:
> > so nobody will try to establish a connection when messenger is stopping?
> 
> I think this implies that each component needs to implement its own `stop()` method properly, to make sure that all the internal continuations/tasks are finished when the future is fulfilled. And it also needs to refuse to process further events from components that are still not shutdown yet. For example, messenger doesn't know that mon client is stopped and will keep dispatching events to it, that's why there are cumbersome `handle_exception_type(...)` logic in this patch.

Yes, I checked the components that OSD contains, it seems that, luckily, only the OSD's stop procedure needs to involve multiple components and I've rearranged it as suggested.

Reply:
I think maybe the handle_exception_type() logic is inevitable, because there maybe continuations that are already submitted to reactors before the stop procedure starts, some of which may fall in some "infinite loop" just like the monc's reconnect operation mentioned above, so we have to notify those continuations to "stop" when doing shutdown, which would inevitable result in an exception being thrown in those continuations that have to be captured by "handle_exception_type()".

Reply:
> > so nobody will try to establish a connection when messenger is stopping?
> 
> I think this implies that each component needs to implement its own `stop()` method properly, to make sure that all the internal continuations/tasks are finished when the future is fulfilled. And it also needs to refuse to process further events from components that are still not shutdown yet. For example, messenger doesn't know that mon client is stopped and will keep dispatching events to it, that's why there are cumbersome `handle_exception_type(...)` logic in this patch.

If I understand correctly, this seems to be a generic problem.
Say, there's a series of continuations "do A; then do B; then do C", if after "do A", the shutdown procedure starts, unless we've kept track of all in-flight continuations, we may inevitably get a error if "do B" or "do C" reference some components that are destroyed in the shutdown procedure. However, is keeping track of all in-flight continuations or unresolved futures really feasible? Am I right about this?

Reply:
> If I understand correctly, this seems to be a generic problem.
Say, there's a series of continuations "do A; then do B; then do C", if after "do A", the shutdown procedure starts, unless we've kept track of all in-flight continuations, we may inevitably get a error if "do B" or "do C" reference some components that are destroyed in the shutdown procedure.

Currently we are already tracking all the in-flight continuations with the help from `ceph::net::Protocol::gated_dispatch()`, none of the components will be destructed before the gate's closing future is fulfilled. That's why our shutdown procedure is "mostly" working. I was thinking whether it could be better if each component can take the responsibility to track their own continuations.

> However, is keeping track of all in-flight continuations or unresolved futures really feasible? Am I right about this?

Messenger is implemented in this way by tracking all the in-flight connections, and all the connections keep track of their own in-flight continuations.

Reply:
Got your point, I'll see what I can do:-)

Reply:
> > so nobody will try to establish a connection when messenger is stopping?
> 
> I think this implies that each component needs to implement its own `stop()` method properly, to make sure that all the internal continuations/tasks are 

i think that's the way to go.

> finished when the future is fulfilled. And it also needs to refuse to process further events from components that are still not shutdown yet. 

i don't really think "refuse" is the right way to describe the behavior, i'd just `assert()` and crash if somebody tries to ring my phone at 4:00 AM.

> For example, messenger doesn't know that mon client is stopped and will keep dispatching events to it, that's why there are cumbersome `handle_exception_type(...)` logic in this patch.

i don't think messenger needs to know this.